로그인/회원가입 API 요청 기능 구현

seongjin0320 opened the pull request:
# 로그인/회원가입 API 요청 기능 구현

## 설명

- 로그인/회원가입 페이지에서 구현되지 않았던 POST API요청 기능을 구현했습니다

- **상세 설명**

- 백엔드 공통 엔드포인트를 환경변수를 통해 주입받아 사용하도록 했습니다. (.env.local)
- axiosInstance를 이용하여 API요청합니다.
- 회원가입시에 넘기는 데이터필드가 백엔드 요구 필드와 맞지 않아 수정하여 회원가입 정상 작동 확인했습니다. 
- 401에러시에 refresh 요청하는 과정을 promise를 변수형으로 유지하지않고 새롭게 생성하여 구독하는 방식으로 변경했습니다. 

- **기타 정보**:
- src/api/endpoints.ts에 엔드포인트 상수화를 했습니다. https://github.com/kakao-tech-campus-3rd-step3/Team8_FE/discussions/53#discussion-8947728
- ~로그인은 관련 이슈가 있어 해결해야할듯 합니다.~ 해결완료
- 중간에 프록시 설정 커밋이 있었는데 그 뒤에 다시 프록시 설정을 삭제했습니다. 주석처리하고싶었으나 충돌방지를 위해 원상태로 돌려놨습니다.
- 에러핸들링은 지금은 최소한의 부분만 해놨고 다음 주차에 표준에러핸들링 로직과 각 페이지에 맞는 세부 에러핸들링을 구현할 예정입니다.

## 관련 이슈(없다면 지워주세요)
- ~회원가입에 성공한 계정으로 로그인 진행 시 400(bad request) 에러가 발생합니다. ~ 해결완료
<!--- 바로 아래에 이슈 링크를 추가해 주세요.-->
https://github.com/kakao-tech-campus-3rd-step3/Team8_FE/issues/52#issue-3455810278 해결완료

seongjin0320 committed:
refact: 인터셉터 내부 리프레시 promise를 새롭게 생성하여 기다리는방식으로 변경

- promise 변수를 생성하여 여러개의 요청들을 묶었던 방식이 JS의 방식과 맞지 않아 제거했습니다
- promise 를 각 요청마다 새롭게 만들어 구독하여 refresh가 끝나기까지 기다리는 방식으로 변경했습니다

seongjin0320 committed:
feat: 로그인/회원가입시에 axiosinstance 호출시에 자동으로 accessToken헤더를 붙이지 않도록 변경

- axiosinstance.ts 의 인터셉터에서 기본적으로 API요청에 accessToken헤더를 붙이도록 되어있는데, 회원가입/로그인 시의 POST요청에 member권한이 필요없으므로 accessToken 헤더를 붙이지 않도록 하는 분기로직을 작성했습니다.
- refresh 요청은 axiosinstance.ts내부에서 axios.post를 호출하므로 axiosinstance.ts 바깥에서 refresh로직을 사용할 일이 없을 것 같아 분기로직에서 생략했습니다.

seongjin0320 committed:
feat: 로그인 API 요청 기능 구현

- useLoginForm.ts를 수정하여 로그인 API 요청 기능을 구현했습니다.
- 로그인과 회원가입 API 요청 엔드포인트를 관리하기 편하게 하기 위해 src/api/endpoints.ts에 엔드포인트를 정의하여 상수화 했습니다.

seongjin0320 committed:
feat: 회원가입 API 요청 기능 구현

- 회원가입 API요청 기능을 로그인 API요청과 일관성 있는 방식으로 axiosInstance.post를 통해 구현했습니다
- 회원가입시에 POST요청으로 보내는 데이터에는 비밀번호 확인 필드 값을 제외하고, mbti는 선택 정보이므로 빈값이면 제거해서 보냅니다.

seongjin0320 committed:
fix: 프록시 설정, 회원가입 request 필드가 서버요구와 달랐던 점 수정

- 서버의 CORS가 해결되지 않아 일시적으로 API 기능 확인을 위해 vite설정을 통해 프록시 설정 했습니다
- 서버의 회원가입 요구 필드와 기존의 회원가입 request 필드가 맞지 않아 400에러가 뜨는 것을 해결했습니다
- 회원가입을 통해 계정을 생성한 뒤, 그 계정으로 로그인 시에 로그인이 안되는 문제가 생겨 해결방법 찾는 중입니다

seongjin0320 committed:
chore: 프록시 설정 원래대로 되돌림

seongjin0320 committed:
fix: 회원가입시에 전화번호 데이터가 - 포함으로 그대로 넘겨지도록 수정

seongjin0320 committed:
refact: AuthContext에서 user정보를 다루지 않고, accessToken, refreshToken만을 다루도록 변경

- /login 요청의 반환값에 user가 없어 로그인시 /me 요청을 통해 user값을 받아와 AuthContext에서 토큰정보와 user데이터를 다뤘었지만, user 정보가 필요한 부분에서 /me 요청을 받는 방식으로 사용하기 위해 AuthContext에서 토큰정보만을 다루도록 변경했습니다
- AuthContext에서 토큰 정보만을 다루기 때문에 useloginForm.ts의 onSubmit시에 로그인 성공 후 login()을 이용해 인증 상태를 관리할 때 토큰 값만 이용하도록 변경했습니다
- 확인용 코드를 통해 실제로 토큰 값이 제대로 저장되고, useAuth를 통해 토큰값에 접근할 수 있음을 확인했습니다

seongjin0320 committed:
feat: 회원가입시에 자동 로그인 기능 추가

Moderator11 commented on src/api/axiosInstance.ts:
```diff
@@ -26,29 +26,57 @@ const clearTokens = () => {
 };
 
+// 로그인/회원가입은 Authorization 헤더 제외 대상 (member권한 필요없음)
+const AUTH_EXCLUDED_PATHS = new Set<string>(['/v1/members/login', '/v1/members/signup']);
```
`endpoint.ts`에도 해주신 것처럼 변하지 않아야 할 객체에는 as const를 붙이면 더 안전합니다!

seongjin0320 replied:
as const를 사용하는 방식으로 변경했습니다~

Moderator11 commented on src/api/axiosInstance.ts:
```diff
@@ -26,29 +26,57 @@ const clearTokens = () => {
 );
 
-// 응답이 401(만료/인증실패)이면 refresh 시도 후 원래 요청 재시도
+// 응답이 401(만료/인증실패)이면 refresh 시도 후 원요청 재시도
```
작은 오타가 잇네용ㅎㅎ 어쩌다가 원요청이 되엇을까요 :)

seongjin0320 replied:
수정완!

Moderator11 commented on src/api/endpoints.ts:
```diff
@@ -0,0 +1,11 @@
+    signup: `${API_PREFIX}/members/signup`,
+    me: `${API_PREFIX}/members/me`,
+  },
+} as const;
```
`as const`👍

Moderator11 commented on src/api/endpoints.ts:
```diff
@@ -0,0 +1,11 @@
+const API_PREFIX = '/v1' as const; //api 프리픽스 수정시 이곳만 수정
```
우왕 API 버전 달라질 것을 대비한 부분 넘 좋슴당

Moderator11 commented on src/pages/register/hooks/useRegisterForm.ts:
```diff
@@ -15,13 +17,24 @@ export const useRegisterForm = () => {
+      // 서버 스펙: { name, contact, email, password, mbti }
+      // confirmPassword 제외, phone -> contact로 매핑(하이픈 제거), mbti 빈 값이면 제외
+      const { confirmPassword, mbti, phone, ...rest } = data;
+      const payload = {
+        ...rest, // name, email, password
+        contact: phone.replace(/-/g, ''),
+        ...(mbti ? { mbti } : {}),
+      };
```
폰 번호를 보낼때 '-' 문자를 없애지 않고 회원가입시 전화번호 유효성 검사 규칙을 바꾸면 어떨까요?

seongjin0320 replied:
하이픈을 없애고 데이터를 넘기는 로직을 실수로 잘못 넣었네요 - 포함 문자열 형태로 보내는것으로 변경했습니다.

Moderator11 commented on src/pages/register/hooks/useRegisterForm.ts:
```diff
@@ -15,13 +17,24 @@ export const useRegisterForm = () => {
+      await axiosInstance.post(ENDPOINTS.members.signup, payload);
+
+      alert('회원가입이 완료되었습니다!');
+      navigate('/login');
```
navigate도 좋지만 라우팅 훅을 만들어 놓앗으니 그걸 사용하면 프로젝트 전체에서 코드 일관성이 높아질 것 가타요

seongjin0320 replied:
라우팅 훅을 사용하도록 변경했습니다

Moderator11 commented on src/api/axiosInstance.ts:
```diff
@@ -26,29 +26,57 @@ const clearTokens = () => {
-let refreshPromise: Promise<string | null> | null = null;
+let refreshSubscribers: Array<(token: string | null) => void> = [];
+
+function subscribeTokenRefresh(cb: (token: string | null) => void) {
+  refreshSubscribers.push(cb);
+}
+
+function notifyTokenRefreshed(token: string | null) {
+  const subs = refreshSubscribers;
+  refreshSubscribers = [];
+  subs.forEach((cb) => {
+    try {
+      cb(token);
+    } catch {}
+  });
+}
```
publish-subscribe 패턴으로 작성하신 것 같은데, 혹시 어떤 동작을 구현하는 것인가요?

seongjin0320 replied:
401에러로 인한(accessToken의 만료) 에러 상황일 때 리프레시토큰을 통해 accessToken을 갱신하는 로직인데, 동시에 401에러가 터졌을 때 각각 refresh를 시키면 문제가 생길 수 있어 제일 먼저 refresh를 시작한 401에 대해서 다른 401에러는 구독자로 등록되어 refresh결과를 기다리게 됩니다. refresh에 성공하면 새 accessToken을 구독자들에게 브로드캐스트 하여 각 원래 요청을 재수행 하게 됩니다. 만약 refresh 를 실패하게 된다면 구독자들에게 실패 신호를 전달하고, 모든 원래 요청들은 에러로 종료됩니다. 그럼 토큰들은 제거가 되고 세션이 해제되어 재 로그인이 필요합니다.

Moderator11 replied:
오홍 상세 설명 감사드립니닷 👍👍

Moderator11 reviewed (approved):
수고하셨습니다 성진님!

저는 이때까지 인증/인가 로직을 axios intercepter로 작성해본 적이 없는데요, 성진님의 코드를 보니 꽤나 딥하게 실패 케이스를 대응하신 것 같아요. publish-subscribe 패턴으로 작성된 부분도 보이는데 잘 이해가 안가네요 😭  이 부분 설명 받고싶어요!

다른 부분들에는 큰 문제는 없는 것 같아서 Approve 남길게욤

seongjin0320 committed:
refact: PR 피드백 반영

- as const를 통해 AUTH_EXCLUDED_PATH 를 정의했습니다
- 사소한 오타 수정 했습니다
- 피드백 내용중에 회원가입시 전화번호 데이터 양식에 대한 부분이 있었는데 이전 커밋에 적용되었습니다
- useLoginForm.ts와 useRegisterForm.ts의 리디렉션 로직을 navigate가 아닌 라우팅 훅을 사용하도록 했습니다
